README.md

```diff
@@ -69,28 +157,98 @@ class AnotherException(Exception):
+    pytest.param(
+        AnotherException('the message'),
+        marks=pytest.mark.raises(exception=AnotherException)
+    ),
+    pytest.param(
+        Exception('the message'),
+        marks=pytest.mark.raises()
+    )
 ])
-def test_mark_raises(error):
+def test_mark_raises_demo(error):
     if error:
         raise error
 
 ```
 
-All of these tests pass.  These examples are actual [tests for this plugin][].
+All of these tests pass.  These examples are actual [tests for this plugin][]
+(exact test case is in `test_pytest_raises_parametrize_demo` test).
+
+### `@pytest.mark.setup_raises` Examples
+
+Usage of the `@pytest.mark.setup_raises` decorator is likely to be uncommon,
+but when it is needed there is no known alternative.  Consider the following
+contrived example, where in a `conftest.py` we have the following check for
+some custom marker we are concerned about:
+
+```python
+# in conftest.py
+def pytest_runtest_setup(item):
+    custom_marker = item.get_closest_marker('custom_marker')
+    if custom_marker:
+        valid = custom_marker.kwargs.get('valid', True)
+        if not valid:
+            raise ValueError('custom_marker.valid was False')
+```
+
+and two tests using this marker
+
+```python
+import pytest
+
+@pytest.mark.custom_marker(valid=False)
+@pytest.mark.setup_raises(
+    exception=ValueError, match=r'.*was False$'
+)
+def test_mark_setup_raises_demo():
+    pass
+
+@pytest.mark.custom_marker(valid=True)
+def test_all_good():
+    pass
+```
+
+This example is in the [tests for this plugin][] in the
+`test_pytest_mark_setup_raises_demo` test case.  This example is awkward, but
+the idea is you can use `@pytest.mark.setup_raises` to catch expected errors
+during the `pytest_runtest_setup` phase.  So when we used `custom_marker`
+with `valid=False`, the `pytest_runtest_setup` will `raise` as expected, but
+not when `valid=True`.
+
+In the real world, the utility of `@pytest.mark.setup_raises` comes in when
+you have potentially less control over the execution of fixtures or perhaps
+want to stress-test custom markers or fixtures.  Consider writing a decorator
+that auto-uses a fixture for a given test function, but deliberately provides
+invalid arguments to the fixture.
+
+In short: the chances are good that you will **not** need
+`@pytest.mark.setup_raises` in the average testing framework.  However, if
+you need to verify failures during the `pytest_runtest_setup` phase, it is
+an invaluable tool.
+
+**Reminder**: notice that when `@pytest.mark.setup_raises` is used, **the
+function body should be exactly `pass`**.  The `pytest_runtest_setup` phase
+has raised, meaning the setup for the test is incomplete.  Anything other
+than an empty test function body of `pass` is **not** supported by this
+extension.
 
 License
 -------
 
-Distributed under the terms of the [MIT][] license, "pytest-raises" is free and open source software
+Distributed under the terms of the [MIT][] license, "pytest-raises" is free and
```

Comment:
This is something I haven't seen before (same with the `[file an issue][]` below), dies github just auto link these? What criteria does it use?

Reply:
Its actually part of the markdown linking/reference standard.
From https://daringfireball.net/projects/markdown/syntax#link

> The implicit link name shortcut allows you to omit the name of the link, in which case the link text itself is used as the name. Just use an empty set of square brackets — e.g., to link the word “Google” to the google.com web site, you could simply write:
> 
> `[Google][]`
> And then define the link:
> 
> `[Google]: http://google.com/`

Reply:
Oh neat, thanks!